File: src/bridge.ts
/**
 * Mnemosyne stdio JSON-RPC bridge for Hermes plugin.
 *
 * Protocol:
 *   → {"id":1,"cmd":"store","args":{"text":"...","opts":{}}}
 *   ← {"id":1,"ok":true,"result":{...}}
 *   → {"id":2,"cmd":"recall","args":{"query":"...","opts":{}}}
 *   ← {"id":2,"ok":true,"result":[...]}
 *   → {"id":3,"cmd":"exit"}
 *   ← (process exits)
 */

import { createMnemosyne, type Mnemosyne } from "./index.js";
import { QdrantDB } from "./core/qdrant.js";
import { DEFAULT_COLLECTIONS } from "./core/types.js";
import { createInterface } from "node:readline";

const QDRANT_URL = process.argv[2] || "http://localhost:6333";
const EMBEDDING_URL = process.argv[3] || "http://localhost:11434/v1/embeddings";
const AGENT_ID = process.argv[4] || "hermes";
const MODEL = process.argv[5] || "mxbai-embed-large";
const COLLECTION_PREFIX = process.argv[6] || "";

let memory: Mnemosyne | null = null;
let ready = false;
let qdrantDb: QdrantDB | null = null;

async function main() {
  const collections = COLLECTION_PREFIX
    ? {
        shared: `${COLLECTION_PREFIX}_shared`,
        private: `${COLLECTION_PREFIX}_private`,
      }
    : undefined;

  memory = await createMnemosyne({
    qdrantUrl: QDRANT_URL,
    embeddingUrl: EMBEDDING_URL,
    agentId: AGENT_ID,
    embeddingModel: MODEL,
    enableDecay: true,
    enableBM25: true,
    enableAutoLink: true,
    collections,
  });
  qdrantDb = new QdrantDB(QDRANT_URL, AGENT_ID, collections);
  ready = true;

  const rl = createInterface({ input: process.stdin, terminal: false });

  for await (const line of rl) {
    const trimmed = line.trim();
    if (!trimmed) continue;

    let req: { id?: number; cmd: string; args?: Record<string, unknown> };
    try {
      req = JSON.parse(trimmed);
    } catch {
      writeErr("invalid JSON");
      continue;
    }

    if (req.cmd === "exit") {
      process.exit(0);
    }

    try {
      await handle(req);
    } catch (err: unknown) {
      const msg = err instanceof Error ? err.message : String(err);
      writeRes(req.id, false, null, msg);
    }
  }
}

async function handle(req: { id?: number; cmd: string; args?: Record<string, unknown> }) {
  if (!memory) {
    writeRes(req.id, false, null, "not initialized");
    return;
  }

  switch (req.cmd) {
    case "ping": {
      writeRes(req.id, true, { ready });
      break;
    }
    case "store": {
      const args = req.args || {};
      const text = args.text as string;
      if (!text) { writeRes(req.id, false, null, "text required"); return; }
      const opts = (args.opts as Record<string, unknown>) || {};
      const result = await memory.store(text, opts);
      writeRes(req.id, true, { id: result });
      break;
    }
    case "recall": {
      const args = req.args || {};
      const query = args.query as string;
      if (!query) { writeRes(req.id, false, null, "query required"); return; }
      const opts = (args.opts as Record<string, unknown>) || {};
      const limit = (opts.limit as number) || 5;
      const minScore = (opts.minScore as number) || 0.3;
      const results = await memory.recall(query, { limit, minScore });
      writeRes(req.id, true, results.map(serializeResult));
      break;
    }
    case "forget": {
      const args = req.args || {};
      const id = args.id as string;
      const query = args.query as string;
      const result = await memory.forget(id || query || "");
      writeRes(req.id, true, { deleted: result });
      break;
    }
    case "toma": {
      // Theory of Mind: query what a SPECIFIC agent knows about a topic
      const args = req.args || {};
      const targetAgent = args.agent_id as string;
      const topic = args.topic as string;
      const limit = (args.limit as number) || 5;
      if (!targetAgent || !topic) {
        writeRes(req.id, false, null, "agent_id and topic required");
        return;
      }
      const sharedColl = COLLECTION_PREFIX
        ? `${COLLECTION_PREFIX}_shared`
        : DEFAULT_COLLECTIONS.SHARED;

      if (!qdrantDb) {
        writeRes(req.id, false, null, "qdrant not initialized");
        return;
      }

      // Embed the query
      const { EmbeddingsClient } = await import("./core/embeddings.js");
      const embedder = new EmbeddingsClient(EMBEDDING_URL, MODEL);
      const vector = await embedder.embed(topic);

      // Search shared collection filtered by agent_id
      const results = await qdrantDb.search(sharedColl, vector, limit, 0.3, {
        agent_id: targetAgent,
      });
      writeRes(req.id, true, results.map(serializeResult));
      break;
    }
    case "profile": {
      // Get agent memory profile: count, top domains, top types
      const args = req.args || {};
      const targetAgent = (args.agent_id as string) || AGENT_ID;
      const sharedColl = COLLECTION_PREFIX
        ? `${COLLECTION_PREFIX}_shared`
        : DEFAULT_COLLECTIONS.SHARED;

      if (!qdrantDb) {
        writeRes(req.id, false, null, "qdrant not initialized");
        return;
      }

      // Count memories for this agent
      const count = await qdrantDb.count(sharedColl);  // Total in collection
      writeRes(req.id, true, {
        agent_id: targetAgent,
        total_memories: count,
      });
      break;
    }
    default: {
      writeRes(req.id, false, null, `unknown cmd: ${req.cmd}`);
    }
  }
}

function serializeResult(r: { entry: { id: string; text: string; memoryType: string; urgency: string; domain: string; confidence: number; importance: number; agentId: string }; score: number }) {
  return {
    id: r.entry.id,
    text: r.entry.text,
    score: r.score,
    memoryType: r.entry.memoryType,
    urgency: r.entry.urgency,
    domain: r.entry.domain,
    confidence: r.entry.confidence,
    importance: r.entry.importance,
    agentId: r.entry.agentId,
  };
}

function writeRes(id: number | undefined, ok: boolean, result: unknown, error?: string) {
  const msg = { id: id ?? 0, ok, result, error: error || null };
  process.stdout.write(JSON.stringify(msg) + "\n");
}

function writeErr(msg: string) {
  process.stderr.write(`[bridge] ${msg}\n`);
}

main().catch((err) => {
  writeErr(String(err));
  process.exit(1);
});

File: plugins/memory/mnemosyne/bridge.py
"""Mnemosyne memory provider — stdio JSON-RPC bridge to TypeScript backend."""

from __future__ import annotations

import fcntl
import json
import logging
import os
import subprocess
import sys
import threading
from typing import Any, Dict, List, Optional

from .cli import get_bridge_command, load_config, is_installed

logger = logging.getLogger(__name__)


class MnemosyneBridge:
    """Manages a long-running Mnemosyne bridge subprocess over stdio JSON-RPC."""

    def __init__(self, hermes_home: str):
        config = load_config(hermes_home)
        self._qdrant_url = config.get("qdrant_url", "http://localhost:6333")
        self._embedding_url = config.get("embedding_url", "http://localhost:11434/v1/embeddings")
        self._model = config.get("embedding_model", "nomic-embed-text")
        self._isolated = config.get("isolated", False)
        self._agent_id = ""
        self._proc: Optional[subprocess.Popen] = None
        self._lock = threading.Lock()
        self._next_id = 1
        self._read_thread: Optional[threading.Thread] = None
        self._responses: Dict[int, threading.Event] = {}
        self._results: Dict[int, Any] = {}
        self._buffer = ""

    def start(self, agent_id: str) -> None:
        """Start the bridge subprocess."""
        self._agent_id = agent_id
        try:
            cmd = get_bridge_command(
                self._qdrant_url,
                self._embedding_url,
                agent_id,
                self._model,
                self._isolated,
            )
        except RuntimeError as e:
            logger.warning("Mnemosyne bridge: %s", e)
            return

        logger.info("Starting Mnemosyne bridge: %s", " ".join(cmd))
        self._proc = subprocess.Popen(
            cmd,
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
        )

        # Start reader thread
        self._read_thread = threading.Thread(target=self._reader, daemon=True)
        self._read_thread.start()

        # Wait for ready signal
        try:
            resp = self._send("ping", timeout=10)
            if resp and resp.get("ready"):
                logger.info("Mnemosyne bridge ready for agent '%s'", agent_id)
            else:
                logger.warning("Mnemosyne bridge responded but not ready")
        except TimeoutError:
            logger.warning("Mnemosyne bridge did not respond to ping within 10s")

    def _send(self, cmd: str, args: Optional[Dict[str, Any]] = None, timeout: float = 30) -> Any:
        """Send a command and wait for response."""
        if not self._proc or not self._proc.stdin:
            raise RuntimeError("Bridge not started")

        with self._lock:
            req_id = self._next_id
            self._next_id += 1
            event = threading.Event()
            self._responses[req_id] = event

        req = json.dumps({"id": req_id, "cmd": cmd, "args": args or {}}) + "\n"
        self._proc.stdin.write(req)
        self._proc.stdin.flush()

        if not event.wait(timeout=timeout):
            with self._lock:
                self._responses.pop(req_id, None)
            raise TimeoutError(f"Bridge command '{cmd}' timed out after {timeout}s")

        with self._lock:
            result = self._results.pop(req_id, None)

        if result is None:
            raise RuntimeError(f"No response for command '{cmd}'")
        if not result.get("ok"):
            raise RuntimeError(result.get("error", f"Command '{cmd}' failed"))
        return result.get("result")

    def _reader(self) -> None:
        """Read JSON lines from bridge stdout and dispatch responses."""
        if not self._proc or not self._proc.stdout:
            return

        # Set stdout to non-blocking
        fd = self._proc.stdout.fileno()
        fl = fcntl.fcntl(fd, fcntl.F_GETFL)
        fcntl.fcntl(fd, fcntl.F_SETFL, fl | os.O_NONBLOCK)

        while True:
            try:
                line = self._proc.stdout.readline()
                if not line:
                    # Check if process died
                    if self._proc.poll() is not None:
                        break
                    continue

                line = line.strip()
                if not line:
                    continue

                try:
                    msg = json.loads(line)
                except json.JSONDecodeError:
                    continue

                req_id = msg.get("id")
                if req_id is not None:
                    with self._lock:
                        event = self._responses.pop(req_id, None)
                        if event:
                            self._results[req_id] = msg
                            event.set()
            except Exception:
                break

        # Process died
        stderr_output = ""
        if self._proc and self._proc.stderr:
            try:
                stderr_output = self._proc.stderr.read()
            except Exception:
                pass
        if stderr_output:
            logger.warning("Mnemosyne bridge stderr: %s", stderr_output.strip())

    def store(self, text: str, opts: Optional[Dict[str, Any]] = None) -> Optional[str]:
        """Store a memory. Returns memory ID or None."""
        result = self._send("store", {"text": text, "opts": opts or {}})
        return result.get("id") if result else None

    def recall(self, query: str, limit: int = 5, min_score: float = 0.3) -> List[Dict[str, Any]]:
        """Recall memories matching query. Returns list of dicts."""
        result = self._send("recall", {"query": query, "opts": {"limit": limit, "minScore": min_score}})
        return result or []

    def forget(self, memory_id: str) -> bool:
        """Forget a memory by ID."""
        result = self._send("forget", {"id": memory_id})
        return result.get("deleted", False) if result else False

    def toma(self, target_agent: str, topic: str, limit: int = 5) -> List[Dict[str, Any]]:
        """Theory of Mind: query what another agent knows about a topic."""
        result = self._send("toma", {
            "agent_id": target_agent,
            "topic": topic,
            "limit": limit,
        })
        return result or []

    def profile(self, target_agent: Optional[str] = None) -> Dict[str, Any]:
        """Get an agent's memory profile."""
        args = {}
        if target_agent:
            args["agent_id"] = target_agent
        result = self._send("profile", args)
        return result or {}

    def ping(self) -> bool:
        """Check if bridge is alive."""
        try:
            result = self._send("ping", timeout=5)
            return result is not None
        except Exception:
            return False

    def shutdown(self) -> None:
        """Shut down the bridge process."""
        try:
            self._send("exit", timeout=2)
        except Exception:
            pass

        if self._proc:
            try:
                self._proc.terminate()
                self._proc.wait(timeout=5)
            except Exception:
                self._proc.kill()
            self._proc = None
